library/Odesk/Phystrix/AbstractCommand.php

```diff
@@ -219,33 +219,32 @@ public function execute()
         if ($this->isRequestCacheEnabled()) {
             $fromCache = $this->requestCache->get($this->getCommandKey(), $this->getCacheKey());
             if ($fromCache !== null) {
-                $this->executionEvents[] = self::EVENT_RESPONSE_FROM_CACHE;
                 $metrics->markResponseFromCache();
+                $this->handleEvent(self::EVENT_RESPONSE_FROM_CACHE);
                 return $fromCache;
             }
         }
         $circuitBreaker = $this->getCircuitBreaker();
         if (!$circuitBreaker->allowRequest()) {
-            $this->executionEvents[] = self::EVENT_SHORT_CIRCUITED;
             $metrics->markShortCircuited();
+            $this->handleEvent(self::EVENT_SHORT_CIRCUITED);
             return $this->getFallbackOrThrowException();
         }
         $this->invocationStartTime = $this->getTimeInMilliseconds();
         try {
             $result = $this->run();
             $this->recordExecutionTime();
-            $this->executionEvents[] = self::EVENT_SUCCESS;
             $metrics->markSuccess();
             $circuitBreaker->markSuccess();
+            $this->handleEvent(self::EVENT_SUCCESS);
         } catch (BadRequestException $exception) {
             // Treated differently and allowed to propagate without any stats tracking or fallback logic
             $this->recordExecutionTime();
             throw $exception;
         } catch (Exception $exception) {
             $this->recordExecutionTime();
-            $this->executionException = $exception;
-            $this->executionEvents[] = self::EVENT_FAILURE;
             $metrics->markFailure();
+            $this->handleEvent(self::EVENT_FAILURE, $exception);
```

Comment:
to insure backwards compatibility and secure internal contract, we could consider calling a private method that will add the event to the list, so this cannot be overridden AND call a protected method, to allow for custom processing of events as they happen.

To think about it, you have access to the list of events already and the list is filled "atomically" with command execution. what is a use case, the need for processing these events as they happen, rather than after execution? @bryanlatten